Aggregate counters

dpino opened the pull request:
Working plan as discussed offline:

- When a worker starts, we start a `recursive_directory_inventory_events()` channel on it to learn about the counters.
- For any file in directory -- say we are opening `/var/run/snabb/1234/` --
- For any file like `/var/run/snabb/foo/bar`, if its basename (lib.basename()) has an extension of ".counter", then it's a counter.
- So we will receive an event on the channel like `{kind='creat', name='/var/run/snabb/1234/foo/bar.counter'}`
- On the manager when we see that, we do `counter.open()` on that counter, store away that counter somewhere and additionally, periodically we create a corresponding counter in the manager process
e.g. /var/run/snabb/5678/foo/bar.counter.
- Where the manager counter = sum of all corresponding worker counters of that name + archived value
as a 64-bit unsigned integer.
- When a counter goes away ({kind='rm', name='/var/run/snabb/1234/foo/bar.counter'}) then we add its last value to the archived counter, then remove it from the set of active worker counters
in that way when we do a "snabb top" on the manager process, we should see a sum of all counters from the workers.

So far I've created the channel and added a fiber for listening the directory events.

dpino committed:
Open channel in worker for listening to directory events

dpino committed:
Initialize aggregated counters in manager process

dpino committed:
Cache counters

dpino committed:
Archive counters on delete

dpino committed:
Move fiber for managing aggregated counters (creation and destruction)

dpino committed:
Create fiber to update aggregated counters

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -306,6 +340,44 @@ function Manager:start_worker_for_graph(id, graph)
+            if event.kind == 'creat' then
+               -- Create aggregated counter in manager process.
+               local aggregated = cname:gsub(worker.pid, S.getpid())
+               if not shm.exists(aggregated) then
```
Probably can move the handling of counter creation to `update_counter`

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -306,6 +340,44 @@ function Manager:start_worker_for_graph(id, graph)
+               end
+            elseif event.kind == 'rm' then
+               -- Move counter from active to archive.
+               archive_counter(cname)
```
Here you need to do `counter.delete` on the worker's counter.  It doesn't actually delete the file, it just unmaps it from memory, for counters opened in a read-only way (via counter.open).  (If it were a counter opened with counter.create it would delete it though.)

dpino committed:
On archiving counter, perform counter.delete

dpino committed:
Merge create and update counter logic

dpino committed:
Sum counters of all workers and commit on each iteration

dpino committed:
Call scheduler shutdown for 0.1s or until it returns true

dpino commented:
The issue with the failing bench test is that apparently schedule.shutdown was returning false. It seemed 100 iterations was not enough. I changed the function to shutdown all tasks until there's no more scheduled. I don't know if that's the right change or just simply should have increased the number of iterations.

wingo commented on src/lib/fibers/sched.lua:
```diff
@@ -88,12 +88,11 @@ function Scheduler:main()
 
 function Scheduler:shutdown()
-   for i=1,100 do
+   repeat
```
I see.  I am not sure about this change.  I agree that we will need to run the scheduler more iterations to drain it, but if something goes wrong while shutting down so that there's an infinite loop and it never finishes, the user should be told that something's wrong.

Can we instead run the scheduler while there are next tasks for up to a certain amount of time (maybe 0.1s), and if after that there are still runnable tasks, warn the user and proceed anyway?  It's not great of course, but maybe everything will just be garbage-collected later.

Besides that -- I think when shutting down we should neatly cancel the inotify streams.  More below.

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -306,6 +340,39 @@ function Manager:start_worker_for_graph(id, graph)
+   fiber.spawn(function ()
+      local worker = self.workers[id]
+      local dir = shm.root..'/'..worker.pid
+      local rx = inotify.recursive_directory_inventory_events(dir)
```
Here you can create a cond to cancel the stream, and use it later.  Like this:

```lua
local cancel = cond.new()
local rx = inotify.recursive_directory_inventory_events(dir, cancel:wait_operation())
...
```
To cancel the stream, do cond:signal().  it will cause "rm" events to be emitted corresponding to all "creat" events so that all counters will be archived and closed.  Probably you'll want to cancel the events when stopping a worker, and also when shutting down the manager.

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -295,6 +297,38 @@ function Manager:compute_scheduling_for_worker(id, app_graph)
+local function archive_counter (name)
+   local c = assert(counters.active[name])
+   counter.delete(c)
+   counters.archived[name] = c
```
To archive a counter, you should sample its current value and add it to the corresponding archived counter.  After `counter.delete`, the memory is no longer mapped, so it is no longer accessible.

dpino committed:
Emit cancel operation on workers shutdown

dpino committed:
Rework archive_counter

wingo commented on src/lib/fibers/sched.lua:
```diff
@@ -92,12 +92,18 @@ function Scheduler:main()
 end
 
 function Scheduler:shutdown()
```
Hi, really sorry for the nit-picking here, but I think this code should be in ptree.  The issue is that the scheduler doesn't always have a notion of time, but when it does, the time to which to advance should be passed as the argument to run; but here in sched.lua we don't really know what to do.  Anyway.  I suggest that ptree should call the old `:shutdown()` in a loop until it returns true or 0.1s have passed.  WDYT?

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -297,17 +301,85 @@ function Manager:compute_scheduling_for_worker(id, app_graph)
+   local val = counter.read(c)
+   counter.delete(name)
+   counters.active[name] = nil
+   counters.archived[name] = counters.archived[name] and counter.set(name, val)
```
many workers can accumulate their values into one archived counter.  e.g. you have 1234/foo/bar.counter and 2345/foo/bar.counter; the aggregate is the sum of all of these, plus the archived value for foo/bar (which can simply be a uint64_t[1] instead of a full counter; probably it doesn't need to go out to the filesystem).  When 1234 goes away you want to add its value to the archived counter.  WDYT?

dpino replied:
OK, I think I understand now better. It's also my understanding that when displaying the aggregated countries the actual value will be the sum of active aggregated counter and archived aggregated "counter".

wingo replied:
yeah i think that's the idea.

dpino committed:
Sum of archived aggregated counter should be accumulative

Also it's not necessary to store the archived value to disk,
storing it in RAM memory is enough.

dpino committed:
Rework aggregate counters logic

Before the aggregated counters were stored in active and its value was
updated without taking into account the archived value.

Now the active table only stores the workers counter indexed by
aggregated counter name.  There is a new table for aggregated counters.
A fiber calculates the accumulative sum of all workers counter than
belong to an aggregated counter and sum the archived value to the
aggregated counter if any.

wingo commented on src/lib/ptree/ptree.lua:
```diff
@@ -297,17 +301,89 @@ function Manager:compute_scheduling_for_worker(id, app_graph)
+-- accumulative sum.  Accumulative sum is initialized to aggregated counter
+-- value if any.
+local function compute_aggregated_value (k)
+   local ret = counters.archived[k] or 0
```
`counters.archived[k][0]` i think

wingo reviewed (approved):
lgtm with one nit

dpino committed:
Fix aggregated counter retrieval
